src/python/pants/backend/java/test/junit.py

```diff
@@ -0,0 +1,125 @@
+from pants.engine.process import FallibleProcessResult, Process
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    CoarsenedTargets,
+    Targets,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.jvm.resolve.coursier_fetch import (
+    CoursierLockfileForTargetRequest,
+    CoursierResolvedLockfile,
+    MaterializedClasspath,
+    MaterializedClasspathRequest,
+    MavenRequirements,
+)
+from pants.jvm.resolve.coursier_setup import Coursier
+from pants.util.logging import LogLevel
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class JavaTestFieldSet(TestFieldSet):
+    required_fields = (JavaTestsSources,)
+
+    sources: JavaTestsSources
+
+
+@rule(desc="Run JUnit", level=LogLevel.DEBUG)
+async def run_junit_test(
+    coursier: Coursier,
+    field_set: JavaTestFieldSet,
+) -> TestResult:
+    transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest([field_set.address]))
+    coarsened_targets = await Get(
+        CoarsenedTargets, Addresses(t.address for t in transitive_targets.closure)
+    )
+    lockfile = await Get(
+        CoursierResolvedLockfile,
+        CoursierLockfileForTargetRequest(Targets(transitive_targets.closure)),
+    )
+    materialized_classpath = await Get(
+        MaterializedClasspath,
+        MaterializedClasspathRequest(
+            prefix="__thirdpartycp",
+            lockfiles=(lockfile,),
+            maven_requirements=(
+                MavenRequirements.create_from_maven_coordinates_fields(
+                    fields=(),
+                    additional_requirements=[
+                        "org.junit.platform:junit-platform-console:1.7.2",
+                        "org.junit.jupiter:junit-jupiter-engine:5.7.2",
+                        "org.junit.vintage:junit-vintage-engine:5.7.2",
+                    ],
+                ),
+            ),
+        ),
+    )
+    transitive_user_classfiles = await MultiGet(
+        Get(CompiledClassfiles, CompileJavaSourceRequest(component=t)) for t in coarsened_targets
+    )
+    merged_transitive_user_classfiles_digest = await Get(
+        Digest, MergeDigests(classfiles.digest for classfiles in transitive_user_classfiles)
+    )
+    usercp_relpath = "__usercp"
+    prefixed_transitive_user_classfiles_digest = await Get(
+        Digest, AddPrefix(merged_transitive_user_classfiles_digest, usercp_relpath)
+    )
+    merged_digest = await Get(
+        Digest,
+        MergeDigests(
+            (
+                prefixed_transitive_user_classfiles_digest,
+                materialized_classpath.digest,
+                coursier.digest,
+            )
+        ),
+    )
+    proc = Process(
+        argv=[
+            coursier.coursier.exe,
+            "java",
+            "--system-jvm",  # TODO(#12293): use a fixed JDK version from a subsystem.
```

Comment:
I assume that this should be fixed to read the option value, and `junit_test.py` should override the option value?

Reply:
Yeah, something to that effect.  It's not entirely obvious to me right now how junit test runner JVM selection should interact with javac JDK selection.  A straightforward next step would be to just make a subsystem like javac's with the same options for JVM version and system JVM special casing, and just consult that here.  Unlike javac, though, junit doesn't need special wrappers around Coursier for finding the literal `javac` binary on the filesystem; it can just use `cs java ...` as seen here.